src/liger_kernel/transformers/model/llama.py

```diff
@@ -247,3 +253,47 @@ def lce_forward(
         hidden_states=outputs.hidden_states,
         attentions=outputs.attentions,
     )
+
+
+def lce_maybe_trainable_lm_head(self, hidden_states, hidden_size, labels, **loss_kwargs):
+    lm_head = self.lm_head
+
+    # Unwrap the module if lm_head has been added as trainable module in PEFT LoRA configuration,
+    # i.e. listed in the modules_to_save field of LoraConfig, so the lm_head weights are read
+    # from the unwrapped module.
+    # See https://huggingface.co/docs/peft/package_reference/lora for reference.
+    if PEFT_AVAILABLE and isinstance(lm_head, ModulesToSaveWrapper):
+        lm_head = lm_head.modules_to_save.default
+
+    # If FSDP is used and lm_head is trainable, e.g., during full fine-tuning or with LoRA,
+    # reading the lm_head module weights and calling the kernel must be done within FSDP forward pass
+    # so the module entire parameters are summoned and kept in memory during the kernel execution.
+    if isinstance(lm_head, FullyShardedDataParallel):
```

Comment:
do you know where does HF explicitly specify `lm_head` to be a leaf FSDP module? last time i checked it only wraps the decoder layers into FSDP 🤔 https://github.com/huggingface/accelerate/blob/b451956fd69a135efc283aadaa478f0d33fcbe6a/src/accelerate/utils/dataclasses.py#L1899C44-L1903 (and that's why directly calling `lm_head` will fail with some tensor storage error because it itself is not a FSDP leaf module thus no FSDP pre-forward will be triggered)

Reply:
I think this is a combination of the accelerate wrapping logic that you've linked plus the one added by PEFT:

https://github.com/huggingface/peft/blob/2d90569c5d264f380ca0df9798c423dc9bf236ff/src/peft/utils/other.py#L939-L985.

The later automatically overwrites accelerate's one when PEFT is used:

https://github.com/huggingface/transformers/blob/b6d65e40b256d98d9621707762b94bc8ad83b7a7/src/transformers/trainer.py#L5223-L5237.

That happens for example when `lm_head` is added to `modules_to_save` as described in https://huggingface.co/docs/trl/sft_trainer#training-adapters.